osutil/writer.go

```diff
@@ -0,0 +1,55 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"fmt"
+)
+
+type BufferLimitError struct{}
+
+func (e BufferLimitError) Error() string {
+	return fmt.Sprintf("buffer capacity exceeded")
+}
+
+type LimitedWriter struct {
+	buf []byte
+}
+
+func NewLimitedWriter(capacity int) *LimitedWriter {
```

Comment:
IMO this should take a `io.Writer` as an argument, so the func bebomes:
```
func NewLimitedWriter(w io.Writer) *LimitedWriter{
    return &LimitedWriter{w}
}
```

Then where you use it:
```
var buf bytes.Buffer
lw := NewLimitedWriter(&buf)
...do something with the writer
```

This way you avoid all the fuss with manually managing the `[]bytes` buffer and so on.

Reply:
Also your `Write()` becomes
```
type LimitedWriter struct{
    w io.Writer
    limit uint64 // limit
    written uint64
}

func (lw *LimitedWriter) Write(data []byte) (int, error) {
    toWrite := len(data)
    leftSpace = lw.limit - lw.written
    if len(data) > leftSpace {
        return 0, fmt.Errorf("not enough space to write %v bytes, capacity is at %v/%v", len(data), lw.written, lw.limit)
    }
    written, err := w.Write(data)
    lw.written += written
    return written, err
}
```

Reply:
BTW. I've implemented something similar right here: https://github.com/mendersoftware/mender/blob/master/utils/limited_writer.go

Reply:
That's a neat suggestion, thanks! Will do that.